### PR TITLE
Sandbox: Remove streaming connections

### PR DESCRIPTION
**Pull request by gerolf-da**
The separate database connection pool for streaming connections
was only used for the active contracts stream. However, a single
db connection was being occupied until the last active contract was
streamed over the Ledger API to the client. This effectively means
that only ever 2 concurrent active contract streams could exist.
No need to say that this is bad design.

The following changes happened:

- remove the db connection pool for streaming connections
- replace the streaming mechanism for active contracts with
  the already existing pagination mechanism in JdbcLedgerDao
- change the pagination mechanism to actually use database level
  limit and offset instead of doing the pagination "client side"
- configure the HikariDataSource with the metric registry

CHANGELOG_BEGIN
- [Sandbox] Improve loading of active contracts for the Sandbox SQL backend.
CHANGELOG_END

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags, if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala [hunk 1]
```diff
@@ -1222,28 +1225,26 @@ private class JdbcLedgerDao(
       .toSet
 
   private val SQL_GET_LEDGER_ENTRIES = SQL(
-    "select * from ledger_entries where ledger_offset>={startInclusive} and ledger_offset<{endExclusive} order by ledger_offset asc")
+    "select * from ledger_entries where ledger_offset>={startInclusive} and ledger_offset<{endExclusive} order by ledger_offset asc limit {pageSize} offset {queryOffset}")
 
   // Note that here we are reading, non transactionally, the stream in chunks. The reason why this is
   // safe is that
   // * The ledger entries are never removed;
   // * We fix the ledger end at the beginning.
```
Comment (SamirTalwar) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala]: This is no longer true. Is it still safe?

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala [hunk 2]
```diff
@@ -1222,28 +1225,26 @@ private class JdbcLedgerDao(
...
   private def paginatingStream[T](
```
Comment (SamirTalwar) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala]: I'd love for this to be pulled out into its own file and wrapped in unit tests, mostly for documentation purposes. It's hard to grok just by reading the code.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala (hunk 1 above)
Comment (gerolf-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala]: start and end are still fixed at the beginning.

Comment (SamirTalwar) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala]: Oh, of course, just not _here_.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/index/JdbcIndexer.scala [hunk 3]
```diff
@@ -28,7 +28,6 @@ import com.digitalasset.platform.sandbox.metrics.timedFuture
 import com.digitalasset.platform.sandbox.stores.ledger.LedgerEntry
 import com.digitalasset.platform.sandbox.stores.ledger.sql.SqlLedger.{
   defaultNumberOfShortLivedConnections,
-  defaultNumberOfStreamingConnections
 }
```
Comment (SamirTalwar) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/index/JdbcIndexer.scala]: We can get rid of the curly braces here.

**Comment by stefanobaghino-da** [comment]
Rebased and force-pushed, let's get this going. :slightly_smiling_face:

**Comment by SamirTalwar** [comment]
Looks like this fails because HikariCP tries to register its metrics twice. Not sure if this is our fault or theirs, but either way, I suggest we revert the metrics stuff in this PR and get it merged, then look at it again.

**Comment by SamirTalwar** [comment]
I'm happy to take this if people agree with the sentiment.

**Comment by gerolf-da** [comment]
@SamirTalwar, @stefanobaghino-da: Thanks for keeping this PR up. I got sidetracked this week. I'll gladly let you take over, @SamirTalwar.

**Comment by gerolf-da** [comment]
This needs a rebase/merge from master, as we have 2 more entry streams now: `JdbcLedgerDao#getConfigurationEntries` and `JdbcLedgerDao#getPartyEntries`.

**Comment by stefanobaghino-da** [comment]
@SamirTalwar I found the problem, we have to reset the metrics registry before each test. I'll override your latest change.

**Comment by gerolf-da** [comment]
Do we not have this problem during recovery as well? We cannot reset the metrics registry in normal operation.
Instead of passing the `MetricRegistry` we could also pass a custom `MetricsTrackerFactory`, that instantiates a modified `CodaHaleMetricsTracker` that calls `getOrAdd` instead of `register` for the gauges.

**Comment by SamirTalwar** [comment]
@stefanobaghino-da: Why not just change `MetricsAround` to use `BeforeAndAfterEach`?

**Comment by SamirTalwar** [comment]
@gerolf-da: I don't think it's a problem in DB connection recovery. I was trying to figure out why we would construct `HikariJdbcConnectionProvider` more than once and I couldn't find anywhere. It's not reconstructed when we recover, so shouldn't re-register the metrics.

**Comment by stefanobaghino-da** [comment]
> @stefanobaghino-da: Why not just change `MetricsAround` to use `BeforeAndAfterEach`?

Makes sense, thanks.

**Comment by stefanobaghino-da** [comment]
The whole things seems quite a concrete jungle of factories, would be probably interesting to rethink a bit the way in which we initialize these after we're done with this.

**Comment by stefanobaghino-da** [comment]
> @stefanobaghino-da: Why not just change `MetricsAround` to use `BeforeAndAfterEach`?

I'm trying this but it makes the problem flaky. Probably some asynchronous behavior going on. I'd keep it as is for now.

**Pull request merged**